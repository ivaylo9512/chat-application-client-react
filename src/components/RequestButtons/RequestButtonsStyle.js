import styled from 'styled-components'

export const Button = styled.button`
    background: rgb(24,27,33);
    margin-left: 1vw;
    height: 2.5vh;
    width: 2.5vh;
    font-size: 1vh;
    margin: 0.5vh 0;
    color: white;
    border-radius: 50%;
    box-shadow: 0px 0px 0.15vw 0.15vw rgb(0 0 0);

    :hover{
        transform: scale(1.1);
    }
`