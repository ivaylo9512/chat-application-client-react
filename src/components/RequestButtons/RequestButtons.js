import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faForward, faCheck, faTimes, faShare, faPaperPlane } from '@fortawesome/free-solid-svg-icons'
import { memo } from 'react';
import { useState } from 'react';
import { setCurrentChat } from 'app/slices/chatsSlice';
import LoadingIndicator from 'components/LoadingIndicator/LoadingIndicator';
import { sendRequest, acceptRequest, denyRequest } from 'app/slices/requestsSlice';
import { useSelector, useDispatch } from 'react-redux';
import { useEffect } from 'react';
import { useHistory } from 'react-router-dom'
import { Button } from 'components/RequestButtons/RequestButtonsStyle';

const RequestButtons = memo(({requestId, userId, initialMessage, chatWithUser}) => {
    const [message, setMessage] = useState(initialMessage);
    const request = useSelector(state => state.requests.data[userId]);
    const [id, setId] = useState(requestId);
    const dispatch = useDispatch();
    const history = useHistory();

    const setChat = () => {
        history.push('/chat');
        dispatch(setCurrentChat(request ? request.chatWithUser : chatWithUser));
    }

    const accept = () => {
        if(!request?.isLoading){
            dispatch(acceptRequest({ id, userId}));
        }
    }

    useEffect(() => {
        if(request && !request.isLoading){
            setMessage(request.state);
            setId(request.id);
        }
    }, [request])

    const getActionButton = () => {
        let action;
        let icon;
        switch(message){
            case 'accept':
                action = () => dispatch(denyRequest({ id, requestState: 'accept', userId }));
                icon = <FontAwesomeIcon icon={faTimes}/>;
                break;
            case 'pending':
                action = () => dispatch(denyRequest({ id, requestState: 'pending', userId }));
                icon = <FontAwesomeIcon icon={faPaperPlane}/>;
                break;
            case 'send':
                action = () => dispatch(sendRequest(userId));
                icon = <FontAwesomeIcon icon={faShare}/>;
                break;
            case 'complete':
                action = setChat;
                icon = <FontAwesomeIcon icon={faForward}/>;
                break;
        }
        
        return(
            <Button data-testid='action' onClick={() => !request?.isLoading && action()}>
                {request?.isLoading
                    ? <LoadingIndicator />
                    : icon
                }
            </Button>
        )
    }

    return(
        <>
            {message == 'accept' &&
                <Button data-testid='accept' onClick={accept}>
                    {request?.isLoading
                        ? <LoadingIndicator />
                        : <FontAwesomeIcon icon={faCheck} />
                    }
                </Button>
            }
            {getActionButton()}
        </>
    )
})
export default RequestButtons