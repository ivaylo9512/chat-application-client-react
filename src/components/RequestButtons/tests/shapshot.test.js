import * as redux from 'react-redux';
import { shallow } from 'enzyme';
import RequestButtons from 'components/RequestButtons/RequestButtons';
import reactRouterDom from 'react-router-dom';

jest.mock('react-router-dom')

const chatWithUser = {id: 1, lastMessage: 'last message', secondUser: { firstName: 'first', lastName: 'last', profileImage: 'image1.png'} };
describe('RequestButtons snapshot tests', () => {

    beforeEach(() => {
        reactRouterDom.useHistory = jest.fn();
        jest.spyOn(redux, 'useDispatch').mockReturnValue(jest.fn());
    })

    const createWrapper = (state, data) => {
        jest.spyOn(redux, 'useSelector').mockReturnValue(state);

        return shallow(
            <RequestButtons requestId={data.requestId} userId={data.userId} initialMessage={data.initialMessage} chatWithUser={data.chatWithUser}/>
        )
    }

    it('should match snapshot with accept state', () => {
        const wrapper = createWrapper(undefined, { requestId: 1, userId: 2, initialMessage: 'accept', chatWithUser: null})

        expect(wrapper).toMatchSnapshot();
    })

    it('should match snapshot with accept state and loading request', () => {
        const wrapper = createWrapper({ id: 2, state: 'accept', isLoading: true }, { requestId: null, userId: 2, initialMessage: 'accept', chatWithUser: null})

        expect(wrapper).toMatchSnapshot();
    })

    it('should match snapshot with complete state', () => {
        const wrapper = createWrapper(undefined, { requestId: 1, userId: 2, initialMessage: 'completed', chatWithUser: null})

        expect(wrapper).toMatchSnapshot();
    })

    it('should match snapshot with send state', () => {
        const wrapper = createWrapper(undefined, { requestId: 1, userId: 2, initialMessage: 'send', chatWithUser: null})

        expect(wrapper).toMatchSnapshot();
    })

    it('should match snapshot with pending state', () => {
        const wrapper = createWrapper(undefined, { requestId: 1, userId: 2, initialMessage: 'pending', chatWithUser: null})

        expect(wrapper).toMatchSnapshot();
    })
})