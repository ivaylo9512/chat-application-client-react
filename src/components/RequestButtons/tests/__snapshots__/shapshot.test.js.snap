// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RequestButtons unit tests should match snapshot with accept state 1`] = `
<Fragment>
  <styled.button
    data-testid="accept"
    onClick={[Function]}
  >
    <FontAwesomeIcon
      border={false}
      className=""
      fixedWidth={false}
      flip={null}
      icon={
        Object {
          "icon": Array [
            512,
            512,
            Array [],
            "f00c",
            "M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z",
          ],
          "iconName": "check",
          "prefix": "fas",
        }
      }
      inverse={false}
      listItem={false}
      mask={null}
      pull={null}
      pulse={false}
      rotation={null}
      size={null}
      spin={false}
      swapOpacity={false}
      symbol={false}
      title=""
      transform={null}
    />
  </styled.button>
  <styled.button
    data-testid="action"
    onClick={[Function]}
  >
    <FontAwesomeIcon
      border={false}
      className=""
      fixedWidth={false}
      flip={null}
      icon={
        Object {
          "icon": Array [
            352,
            512,
            Array [],
            "f00d",
            "M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z",
          ],
          "iconName": "times",
          "prefix": "fas",
        }
      }
      inverse={false}
      listItem={false}
      mask={null}
      pull={null}
      pulse={false}
      rotation={null}
      size={null}
      spin={false}
      swapOpacity={false}
      symbol={false}
      title=""
      transform={null}
    />
  </styled.button>
</Fragment>
`;

exports[`RequestButtons unit tests should match snapshot with accept state and loading request 1`] = `
<Fragment>
  <styled.button
    data-testid="accept"
    onClick={[Function]}
  >
    <LoadingIndicator />
  </styled.button>
  <styled.button
    data-testid="action"
    onClick={[Function]}
  >
    <LoadingIndicator />
  </styled.button>
</Fragment>
`;

exports[`RequestButtons unit tests should match snapshot with complete state 1`] = `
<Fragment>
  <styled.button
    data-testid="action"
    onClick={[Function]}
  >
    <FontAwesomeIcon
      border={false}
      className=""
      fixedWidth={false}
      flip={null}
      icon={
        Object {
          "icon": Array [
            512,
            512,
            Array [],
            "f04e",
            "M500.5 231.4l-192-160C287.9 54.3 256 68.6 256 96v320c0 27.4 31.9 41.8 52.5 24.6l192-160c15.3-12.8 15.3-36.4 0-49.2zm-256 0l-192-160C31.9 54.3 0 68.6 0 96v320c0 27.4 31.9 41.8 52.5 24.6l192-160c15.3-12.8 15.3-36.4 0-49.2z",
          ],
          "iconName": "forward",
          "prefix": "fas",
        }
      }
      inverse={false}
      listItem={false}
      mask={null}
      pull={null}
      pulse={false}
      rotation={null}
      size={null}
      spin={false}
      swapOpacity={false}
      symbol={false}
      title=""
      transform={null}
    />
  </styled.button>
</Fragment>
`;

exports[`RequestButtons unit tests should match snapshot with pending state 1`] = `
<Fragment>
  <styled.button
    data-testid="action"
    onClick={[Function]}
  >
    <FontAwesomeIcon
      border={false}
      className=""
      fixedWidth={false}
      flip={null}
      icon={
        Object {
          "icon": Array [
            512,
            512,
            Array [],
            "f1d8",
            "M476 3.2L12.5 270.6c-18.1 10.4-15.8 35.6 2.2 43.2L121 358.4l287.3-253.2c5.5-4.9 13.3 2.6 8.6 8.3L176 407v80.5c0 23.6 28.5 32.9 42.5 15.8L282 426l124.6 52.2c14.2 6 30.4-2.9 33-18.2l72-432C515 7.8 493.3-6.8 476 3.2z",
          ],
          "iconName": "paper-plane",
          "prefix": "fas",
        }
      }
      inverse={false}
      listItem={false}
      mask={null}
      pull={null}
      pulse={false}
      rotation={null}
      size={null}
      spin={false}
      swapOpacity={false}
      symbol={false}
      title=""
      transform={null}
    />
  </styled.button>
</Fragment>
`;

exports[`RequestButtons unit tests should match snapshot with send state 1`] = `
<Fragment>
  <styled.button
    data-testid="action"
    onClick={[Function]}
  >
    <FontAwesomeIcon
      border={false}
      className=""
      fixedWidth={false}
      flip={null}
      icon={
        Object {
          "icon": Array [
            512,
            512,
            Array [],
            "f064",
            "M503.691 189.836L327.687 37.851C312.281 24.546 288 35.347 288 56.015v80.053C127.371 137.907 0 170.1 0 322.326c0 61.441 39.581 122.309 83.333 154.132 13.653 9.931 33.111-2.533 28.077-18.631C66.066 312.814 132.917 274.316 288 272.085V360c0 20.7 24.3 31.453 39.687 18.164l176.004-152c11.071-9.562 11.086-26.753 0-36.328z",
          ],
          "iconName": "share",
          "prefix": "fas",
        }
      }
      inverse={false}
      listItem={false}
      mask={null}
      pull={null}
      pulse={false}
      rotation={null}
      size={null}
      spin={false}
      swapOpacity={false}
      symbol={false}
      title=""
      transform={null}
    />
  </styled.button>
</Fragment>
`;
